### PR TITLE
Create Clusters Api

### PR DESCRIPTION
**Pull request by arunvasudevan**
# stream-registry PR

Create /Clusters Api

### Added
/Clusters as First class citizen with put(), get(clusterName), getAll()

# PR Checklist Forms

- [X] CHANGELOG.md updated
- [X] Reviewer assigned
- [X] PR assigned (presumably to submitter)
- [X] Labels added (enhancement, bug, documentation)

### CODE REVIEW

**Comment by arunvasudevan** [comment]
Closes #36

File: core/src/main/java/com/homeaway/streamplatform/streamregistry/db/dao/AbstractDao.java [hunk 1]
```diff
@@ -64,59 +62,32 @@
...
+        
         this.kafkaProducer = managedKafkaProducer;
         this.kStreams = kStreams;
         this.env = env;
         this.regionDao = regionDao;
+        this.clusterDao = clusterDao;
         this.infraManager = infraManager;
     }
 
-    /**
```
Comment (arunvasudevan) [inline core/src/main/java/com/homeaway/streamplatform/streamregistry/db/dao/AbstractDao.java]: Moved getClusterDetails() to ClusterDaoImpl method

File: core/src/main/java/com/homeaway/streamplatform/streamregistry/db/dao/impl/ClusterDaoImpl.java [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        }
+
+        return Optional.of(AvroToJsonDTO.getJsonClusterValue(clustersByName.get(clusterName)));
+    }
+
+    @Override
+    public void upsertCluster(JsonCluster jsonCluster) throws InvalidClusterException {
+        log.info("Upserting Cluster {}", jsonCluster);
+
+        ClusterValidator.isValid(jsonCluster);
```
Comment (arunvasudevan) [inline core/src/main/java/com/homeaway/streamplatform/streamregistry/db/dao/impl/ClusterDaoImpl.java]: Validating the cluster object and throwing InvalidCluster Exception before upsert

File: core/src/main/java/com/homeaway/streamplatform/streamregistry/model/ClusterValue.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+import lombok.ToString;
+
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.fasterxml.jackson.databind.annotation.JsonPOJOBuilder;
+
+@JsonDeserialize(builder = ClusterValue.ClusterValueBuilder.class)
+@Builder
+@Getter
+@ToString
+public class ClusterValue {
```
Comment (arunvasudevan) [inline core/src/main/java/com/homeaway/streamplatform/streamregistry/model/ClusterValue.java]: Creates as concrete fields instead of a map to be helpful for users in the swagger

File: core/src/main/java/com/homeaway/streamplatform/streamregistry/resource/ClusterResource.java [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+        value = "Get a Cluster",
+        notes = "Returns a single cluster",
+        tags = "clusters",
+        response = ClusterValue.class)
+    @ApiResponses(value = { @ApiResponse(code = 200, message = "Returns Cluster information", response = ClusterValue.class),
+        @ApiResponse(code = 500, message = "Error Occurred while getting data"),
+        @ApiResponse(code = 404, message = "Cluster not found") })
+    @Produces(MediaType.APPLICATION_JSON)
+    @Timed
+    public Response getCluster(@ApiParam(value = "clusterName", required = true) @PathParam("clusterName") String clusterName) {
```
Comment (arunvasudevan) [inline core/src/main/java/com/homeaway/streamplatform/streamregistry/resource/ClusterResource.java]: Get by clustername - ClusterName is currently not part of the key

File: infra-provider/src/main/java/com/homeaway/streamplatform/streamregistry/provider/InfraManager.java [hunk 5]
```diff
@@ -26,25 +26,37 @@
...
  * This map will be returned to a Stream Registry client at the time of registration.<br>
  * <br>
  * An example of a Confluent Platform Kafka Cluster would contain the following details
  * <ul>
- *     <li><code>bootstrap.servers</code></li>
- *     <li><code>schema.registry.url</code></li>
+ * <li><code>bootstrap.servers</code></li>
+ * <li><code>schema.registry.url</code></li>
+ * <li><code>cluster.name</code></li>
+ * <li><code>zookeeper.quorum</code></li>
```
Comment (arunvasudevan) [inline infra-provider/src/main/java/com/homeaway/streamplatform/streamregistry/provider/InfraManager.java]: Added missing clusterValues

File: core/src/main/java/com/homeaway/streamplatform/streamregistry/db/dao/ClusterDao.java [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+import com.homeaway.streamplatform.streamregistry.exceptions.ClusterNotFoundException;
+import com.homeaway.streamplatform.streamregistry.model.ClusterKey;
+import com.homeaway.streamplatform.streamregistry.model.ClusterValue;
+import com.homeaway.streamplatform.streamregistry.model.JsonCluster;
+
+/**
+ * Interface for Cluster dao.
+ */
+public interface ClusterDao {
+    Map<ClusterKey, ClusterValue> getAllClusters();
```
Comment (prabhakar1983) [inline core/src/main/java/com/homeaway/streamplatform/streamregistry/db/dao/ClusterDao.java]: can we add java docs please

File: core/src/main/java/com/homeaway/streamplatform/streamregistry/exceptions/InvalidClusterException.java [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+ * limitations under the License.
+ */
+package com.homeaway.streamplatform.streamregistry.exceptions;
+
+import com.homeaway.streamplatform.streamregistry.model.JsonCluster;
+
+public class InvalidClusterException extends RuntimeException{
+
+    protected final JsonCluster cluster;
+
```
Comment (prabhakar1983) [inline core/src/main/java/com/homeaway/streamplatform/streamregistry/exceptions/InvalidClusterException.java]: why should have a constructor with no argument for message ?

File: core/src/main/java/com/homeaway/streamplatform/streamregistry/extensions/validation/ClusterValidator.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+package com.homeaway.streamplatform.streamregistry.extensions.validation;
+
+import com.homeaway.streamplatform.streamregistry.exceptions.InvalidClusterException;
+import com.homeaway.streamplatform.streamregistry.model.ClusterKey;
+import com.homeaway.streamplatform.streamregistry.model.ClusterValue;
+import com.homeaway.streamplatform.streamregistry.model.JsonCluster;
+
+public class ClusterValidator {
+
+    public static boolean isValid(JsonCluster cluster) throws InvalidClusterException {
```
Comment (prabhakar1983) [inline core/src/main/java/com/homeaway/streamplatform/streamregistry/extensions/validation/ClusterValidator.java]: can we the method name to 
```
public static boolean isValid(JsonCluster cluster) throws InvalidClusterException
```
so that an exception will be thrown in case of failure.

File: core/src/main/java/com/homeaway/streamplatform/streamregistry/exceptions/InvalidClusterException.java (hunk 7 above)
Comment (prabhakar1983) [inline core/src/main/java/com/homeaway/streamplatform/streamregistry/exceptions/InvalidClusterException.java]: just a opinion. we can extend Exception, so that it has to be thrown and mentioned in the Interface as well.
in that way, client would know that they have to take action on this exception.

File: core/src/main/java/com/homeaway/streamplatform/streamregistry/resource/ClusterResource.java [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+            log.info("Cluster Object {}", clusterParam);
+
+            clusterDao.upsertCluster(clusterParam);
+            return Response.status(Response.Status.ACCEPTED).build();
+        } catch (RuntimeException e) {
+            log.error("Error upserting cluster.", e);
+            return Response.status(Response.Status.BAD_REQUEST)
+                .entity(new ErrorMessage(Response.Status.BAD_REQUEST.getStatusCode(),
+                    "Error upserting cluster.",
+                    e.getCause() != null ? e.getMessage() + ". " + e.getCause().getMessage() : e.getMessage()))
```
Comment (prabhakar1983) [inline core/src/main/java/com/homeaway/streamplatform/streamregistry/resource/ClusterResource.java]: can we add atleast the key for the error message
`"Error upserting cluster."`

File: core/src/main/java/com/homeaway/streamplatform/streamregistry/resource/ClusterResource.java [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+
+            if(type != null && !type.isEmpty()) {
+                clusterMap = clusterMap.entrySet().stream()
+                    .filter(e -> e.getKey().getType().equalsIgnoreCase(type))
+                    .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+            }
+
+            return Response.status(200).entity(clusterMap).build();
+        } catch (Exception e) {
+            log.error("Error getting cluster details.", e);
```
Comment (prabhakar1983) [inline core/src/main/java/com/homeaway/streamplatform/streamregistry/resource/ClusterResource.java]: lets catch any runtime exception if any.

if it is checked exceptions, lets catch the particular ones and take a action on it

File: core/src/test/java/com/homeaway/streamplatform/streamregistry/db/dao/impl/ClusterDaoImplTest.java [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+        // Change the cluster properties to make the assert fail
+        Map<String, String> newClusterProperties = new HashMap<>();
+        newClusterProperties.put(CLUSTER_NAME, "cluster_dao_test_primary_notwork");
+        actualClusterValue.setClusterProperties(newClusterProperties);
+
+        Assert.assertNotEquals(expectedClusterValue, actualClusterValue);
+    }
+
+    /**
+     * Test get all clusters.
```
Comment (prabhakar1983) [inline core/src/test/java/com/homeaway/streamplatform/streamregistry/db/dao/impl/ClusterDaoImplTest.java]: 👍

File: core/pom.xml [hunk 12]
```diff
@@ -162,7 +162,7 @@
 
         <!-- Swagger -->
         <dependency>
-            <groupId>io.swagger</groupId>
+            <groupId>io.swagger.core.v3</groupId>
```
Comment (arunvasudevan) [inline core/pom.xml]: Updating swagger annotations

Comment (neoword) [reply core/pom.xml]: Why do we need this updated for this PR?
Consider doing this as a separate PR.

**Comment by cricket007** [comment]
Why are you trying to merge in commits from December? Please rebase.

File: core/src/main/java/com/homeaway/streamplatform/streamregistry/db/dao/ClusterDao.java [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+
+    com.homeaway.digitalplatform.streamregistry.ClusterValue getCluster(String vpc, String env, String hint, String actorType) throws ClusterNotFoundException;
```
[the remark this answers is not in the document]
Comment (arunvasudevan) [reply core/src/main/java/com/homeaway/streamplatform/streamregistry/db/dao/ClusterDao.java]: One is used internally by `streamDao` to get an `AvroCluster` the other is called by the resource to get the Json Deserialized Cluster using Clustername (not in key).

File: core/src/main/java/com/homeaway/streamplatform/streamregistry/db/dao/ClusterDao.java [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+
+    Optional<ClusterValue> getCluster(String clusterName);
+
+    void upsertCluster(JsonCluster jsonCluster);
```
[the remark this answers is not in the document]
Comment (arunvasudevan) [reply core/src/main/java/com/homeaway/streamplatform/streamregistry/db/dao/ClusterDao.java]: Why separate insert/update? Are we planning the same for streams/producers/consumers?

[the remark this answers is not in the document]
Comment (arunvasudevan) [reply core/src/main/java/com/homeaway/streamplatform/streamregistry/db/dao/ClusterDao.java]: Exposing delete in the Api without any authentication is risky. Application completely depends on entries in these clusters if someone accidently deletes it, it could cause broken streams

File: core/src/main/java/com/homeaway/streamplatform/streamregistry/db/dao/ClusterDao.java [hunk 15]
```diff
@@ -0,0 +1,37 @@
+/* Copyright (c) 2018 Expedia Group.
```
[the remark this answers is not in the document]
Comment (arunvasudevan) [reply core/src/main/java/com/homeaway/streamplatform/streamregistry/db/dao/ClusterDao.java]: When i change the license.txt its touching all the files. Let me do it as a separate PR

File: core/src/main/java/com/homeaway/streamplatform/streamregistry/db/dao/ClusterDao.java (hunk 6 above)
Comment (arunvasudevan) [reply core/src/main/java/com/homeaway/streamplatform/streamregistry/db/dao/ClusterDao.java]: Added

File: core/src/main/java/com/homeaway/streamplatform/streamregistry/resource/ClusterResource.java (hunk 9 above)
Comment (arunvasudevan) [reply core/src/main/java/com/homeaway/streamplatform/streamregistry/resource/ClusterResource.java]: Added more details in the message when possible.

Comment (prabhakar1983) [reply core/src/main/java/com/homeaway/streamplatform/streamregistry/resource/ClusterResource.java]: i may put 
```
new ErrorMessage(Response.Status.BAD_REQUEST.getStatusCode(),
                    String.format(""Error upserting cluster:%s. %s", clusterParam, e.getMessage()),
                    e.getCause() != null ? e.getCause().getMessage() : null))
```

File: core/src/main/java/com/homeaway/streamplatform/streamregistry/resource/ClusterResource.java (hunk 4 above)
Comment (prabhakar1983) [reply core/src/main/java/com/homeaway/streamplatform/streamregistry/resource/ClusterResource.java]: however, cant we just add `clusterName` as a query param in `getAllClusters` , and do the same filtering that has been done in this method.

File: core/src/main/java/com/homeaway/streamplatform/streamregistry/db/dao/ClusterDao.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (arunvasudevan) [reply core/src/main/java/com/homeaway/streamplatform/streamregistry/db/dao/ClusterDao.java]: Removed one of the getCluster() instead used `filter` to map by `clusterName`

File: infra-provider-kafka/pom.xml [hunk 16]
```diff
@@ -46,10 +46,15 @@
...
         </dependency>
+
         <dependency>
             <groupId>org.apache.commons</groupId>
             <artifactId>commons-lang3</artifactId>
         </dependency>
+        <dependency>
+            <groupId>io.confluent</groupId>
+            <artifactId>kafka-avro-serializer</artifactId>
+        </dependency>
```
Comment (neoword) [inline infra-provider-kafka/pom.xml]: why was this necessary on this PR (why didn't we have this already?)
